### DISCUSSION TITLE
Generated query joins the same table twice

### DISCUSSION THREAD

**Discussion by dbettini** [mikro-orm/mikro-orm#3963] [Q&A]
**Describe the bug**
When trying to fetch all entities of a specific base class and populate the relations that are defined on the child classes, the generated query joins the relation tables twice. I have bumped into this problem after adding another relation to the join and getting a stack overflow each time this query is done after the `related` table got to 80 records, which is very low.
This of course only happens using the JOINED loading strategy, which is set globally.

**To Reproduce**

```ts
@Entity({
  abstract: true,
  discriminatorColumn: 'type'
})
class Base {
	type: string;
}

@Entity({
  discriminatorValue: 'first'
})
class DerivedFirst extends Base {
	@OneToMany(() => RelatedFirst, r => r.derived)
  	relatedFirstItems = new Collection<RelatedFirst>(this);
}

@Entity({
  discriminatorValue: 'second'
})
class DerivedSecond extends Base {
	@OneToMany(() => RelatedSecond, r => r.derived)
	relatedSecondItems = new Collection<RelatedSecond>(this);
}

@Entity({
  abstract: true,
  discriminatorColumn: 'type'
})
class Related {
	type: string;
}

@Entity({
  discriminatorValue: 'first'
})
class RelatedFirst extends Related {
  	@ManyToOne(() => DerivedFirst)
  	derived: DerivedFirst;
}

@Entity({
  discriminatorValue: 'second'
})
class RelatedSecond extends Related  {
  	@ManyToOne(() => DerivedSecond)
  	derived: DerivedSecond;
}
...
const entities = await baseRepository.findAll(['relatedFirstItems', 'relatedSecondItems']);
```

**Expected behavior**
Getting an array of all instances derived from Base, casted into their derived classes, with their relations loaded. This *does* happen, but the `related` table is joined twice.

**Additional context**
If this was solved in a newer version, please let me know, but I wasn't able to find it. The way I went around this was just doing the population of other relations afterwards, so they're not added to this join, but it still seems this could be dangerous. 
Another approach, of course, would be to use SELECT_IN load strategy.
Could it be that this approach is wrong in the first place? There's probably no type safe way to do this by using the object version of `populate`. Is the recommended way maybe just fetching each entity separately (even if they are in the same table and subclass the same base), and then merging the arrays manually?

**Versions**
4.5.7

| Dependency | Version |
| - | - |
| node | 14.20.0 |
| typescript | 4.3.5 |
| mikro-orm | 4.5.7 |
| your-driver | Postgres |

**Comment by B4nan** [comment]
This is expected behavior, you always get a single join for each populated relation, even if they target the same table (even without STI). I'll be happy to look into a complete repro if you are facing some perf issues. Please keep in mind you are using a very outdated version, there won't be any updates to v4 branch, and the first thing to actually do is to check if this is a problem in v5 (but the behavior is for sure the same - one join per one relation).
